contracts/StableJoeStaking.sol

```diff
@@ -0,0 +1,181 @@
+import "@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/SafeERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+/**
+ * @title Stable Joe Staking
+ * @author traderjoexyz
+ * @notice StableJoeStaking is a contract that allows to deposit moJoe and receives stable coin sent by JoeMakerV4's daily
+ * harvests. Users deposit moJoe and receive a share of what has been sent by JoeMakerV4 based on their participation
+ * of the total deposited moJoe.
+ */
+contract StableJoeStaking is Initializable, OwnableUpgradeable {
+    using SafeMathUpgradeable for uint256;
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    /// @notice Info of each user.
+    struct UserInfo {
+        uint256 amount; /// @notice How many LP rewardTokens the user has provided.
+        uint256 rewardDebt; /// @notice The sum of rewards already paid to this account. See explanation below.
+        /**
+         * @notice We do some fancy math here. Basically, any point in time, the amount of JOEs
+         * entitled to a user but is pending to be distributed is:
+         *
+         *   pending reward = (user.amount * accTokenPerShare) - user.rewardDebt
+         *
+         * Whenever a user deposits or withdraws moJoe. Here's what happens:
+         *   1. accTokenPerShare (and `lastTokenBalance`) gets updated.
+         *   2. User receives the pending reward sent to his/her address.
+         *   3. User's `amount` gets updated.
+         *   4. User's `rewardDebt` gets updated.
+         */
+    }
+
+    /// @notice Info.
+    IERC20Upgradeable moJoe; /// @notice Address of moJoe contract.
+    uint256 lastTokenBalance; /// @notice Last balance of reward token.
+    uint256 accTokenPerShare; /// @notice Accumulated Tokens per share, times 1e12. See above.
+
+    /// @notice The reward token.
+    IERC20Upgradeable public rewardToken;
+
+    /// @notice Info of each user that stakes moJoe.
+    mapping(address => UserInfo) public userInfo;
+
+    /// @notice Emitted when a user deposits its moJOE
+    event Deposit(address indexed user, uint256 amount);
+
+    /// @notice Emitted when a user withdraws its moJOE
+    event Withdraw(address indexed user, uint256 amount);
+
+    /// @notice Emitted when a user emergency withdraws its moJOE
+    event EmergencyWithdraw(address indexed user, uint256 amount);
+
+    /**
+     * @notice Initialize a new StableJoeStaking contract.
+     * @dev This contract needs to receive an ERC20 `_rewardToken` in order to distribute them
+     * (with JoeMakerV4 in our case)
+     * @param _rewardToken The address of the ERC-20 reward token
+     * @param _moJoe The address of the ERC-20 moJOE token, traderjoe's staking token.
+     */
+    function initialize(IERC20Upgradeable _rewardToken, IERC20Upgradeable _moJoe) public initializer {
```

Comment:
nit: mark as `external`, unless theres a reason `initialize` needs to be `public`